apps/emqx_resource/src/emqx_resource_manager.erl

```diff
@@ -0,0 +1,408 @@
+            ets:insert(?ETS_TABLE, {Data#data.id, Data#data.group, UpdatedData}),
+            {next_state, connected, UpdatedData};
+        {error, Reason} ->
+            logger:error("health check for ~p failed: ~p", [Data#data.id, Reason]),
+            UpdatedData = Data#data{error = Reason},
+            Actions = [{state_timeout, ?SHORT_HEALTHCHECK_INTERVAL, healthcheck}],
+            {keep_state, UpdatedData, Actions};
+        {error, Reason, ResourceState} ->
+            logger:error("health check for ~p failed: ~p", [Data#data.id, Reason]),
+            UpdatedData = Data#data{state = ResourceState, error = Reason},
+            Actions = [{state_timeout, ?SHORT_HEALTHCHECK_INTERVAL, healthcheck}],
+            {keep_state, UpdatedData, Actions}
+    end.
+
+perform_connected_healthcheck(Data) ->
+    case emqx_resource:call_health_check(Data#data.id, Data#data.mod, Data#data.state) of
+        {ok, ResourceState} ->
+            UpdatedData = Data#data{state = ResourceState},
+            ets:insert(?ETS_TABLE, {Data#data.id, Data#data.group, UpdatedData}),
+            Actions = [{state_timeout, ?HEALTHCHECK_INTERVAL, healthcheck}],
+            {keep_state, UpdatedData, Actions};
+        {error, Reason} ->
+            logger:error("health check for ~p failed: ~p", [Data#data.id, Reason]),
+            UpdatedData = Data#data{error = Reason},
+            ets:delete(?ETS_TABLE, Data#data.id),
+            {next_state, connecting, UpdatedData};
+        {error, Reason, ResourceState} ->
+            logger:error("health check for ~p failed: ~p", [Data#data.id, Reason]),
+            UpdatedData = Data#data{state = ResourceState, error = Reason},
+            ets:delete(?ETS_TABLE, Data#data.id),
+            {next_state, connecting, UpdatedData}
+    end.
+
+data_record_to_external_map(Data) ->
+    #{
+        id => Data#data.id,
+        mod => Data#data.mod,
+        config => Data#data.config,
+        status => Data#data.status,
+        state => Data#data.state
+    }.
+
+data_record_to_external_map_with_metrics(Data) ->
+    DataMap = data_record_to_external_map(Data),
+    DataMap#{metrics => get_metrics(Data#data.id)}.
+
+make_test_id() ->
+    RandId = iolist_to_binary(emqx_misc:gen_id(16)),
+    <<?TEST_ID_PREFIX, RandId/binary>>.
+
+wait_for_resource_ready(InstId, WaitTime) ->
+    do_wait_for_resource_ready(InstId, WaitTime div 100).
+
+do_wait_for_resource_ready(_InstId, 0) ->
+    timeout;
+do_wait_for_resource_ready(InstId, Retry) ->
+    case lookup(InstId) of
+        {ok, _Group, #{status := connected}} ->
+            ok;
+        _ ->
+            timer:sleep(100),
+            do_wait_for_resource_ready(InstId, Retry - 1)
+    end.
+
+%% @doc Get the metrics for the specified resource
+get_metrics(InstId) ->
+    emqx_plugin_libs_metrics:get_metrics(resource_metrics, InstId).
+
+safe_call(InstId, Message) ->
+    try
+        gen_statem:call(proc_name(InstId), Message)
+    catch
+        exit:_ ->
```

Comment:
Suggestion: maybe assert that the exit is due do `noproc` to avoid capturing other errors and reporting `not_found`?

```suggestion
        exit:{noproc, _} ->
```